cpp/src/arrow/record_batch.h

```diff
@@ -227,6 +227,11 @@ class ARROW_EXPORT RecordBatchReader {
   /// \return Status
   virtual Status ReadNext(std::shared_ptr<RecordBatch>* batch) = 0;
 
+  virtual Status ReadNext(std::shared_ptr<RecordBatch>* batch,
+                          std::shared_ptr<KeyValueMetadata>* custom_metadata) {
+    return Status::NotImplemented("ReadNext with custom metadata");
+  }
```

Comment:
I add this new API with a default implementation so that existing sub class of `RecordBatchReader` can still be compiled successfully.

Reply:
Hmm, should we prefer `arrow::Result` for new APIs?

Reply:
Perhaps something like:
```c++
struct RecordBatchWithMetadata {
  std::shared_ptr<RecordBatch> batch;
  std::shared_ptr<KeyValueMetadata> custom_metadata;
};
virtual Result<RecordBatchWithMetadata> ReadNext();
```

Reply:
I change the code to return `Result<RecordBatchWithMetadata>` now.